### PR TITLE
EVP_CIPHER_CTX_get_key_length and EVP_CIPHER_CTX_get_iv_length speedup

### PR DESCRIPTION
**Pull request by t8m**
This is just a draft PR for the approach I've suggested. IMO this does not really complicate the get_ctx_param code however of course it is not really generic approach and speeds up in practice only the case where we get a single param and one that is near the beginning of the params we check in the get_ctx_param code.

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by t8m** [comment]
Some numbers for the speedup measured by the openssl speed:

Original:
```
[mraz@leccinum ossl2]$ LD_LIBRARY_PATH=. apps/openssl speed -seconds 1 -evp aes-128-gcm
Doing AES-128-GCM for 1s on 16 size blocks: 40193219 AES-128-GCM's in 1.00s
Doing AES-128-GCM for 1s on 64 size blocks: 19933155 AES-128-GCM's in 1.00s
Doing AES-128-GCM for 1s on 256 size blocks: 12005343 AES-128-GCM's in 1.00s
Doing AES-128-GCM for 1s on 1024 size blocks: 4264277 AES-128-GCM's in 1.00s
Doing AES-128-GCM for 1s on 8192 size blocks: 609969 AES-128-GCM's in 1.00s
Doing AES-128-GCM for 1s on 16384 size blocks: 308704 AES-128-GCM's in 1.00s
Doing get-info for 1s on 16 size blocks: 11193804 get-info's in 1.00s
version: 3.1.0-dev
built on: Wed Dec  8 17:45:44 2021 UTC
options: bn(64,64)
compiler: gcc -fPIC -pthread -m64 -Wa,--noexecstack -Wall -O3 -DPEDANTIC -pedantic -Wno-long-long -DUNUSEDRESULT_DEBUG -Wall -Wmissing-declarations -Wextra
 -Wno-unused-parameter -Wno-missing-field-initializers -Wswitch -Wsign-compare -Wshadow -Wformat -Wtype-limits -Wundef -Werror -Wmissing-prototypes -Wstric
t-prototypes -g -O2 -Werror -DOPENSSL_USE_NODELETE -DL_ENDIAN -DOPENSSL_PIC -DOPENSSL_BUILDING_OPENSSL -DZLIB -DNDEBUG
CPUINFO: OPENSSL_ia32cap=0x7ed8320b078bffff:0x400004219c91a9
The 'numbers' are in 1000s of bytes per second processed.
type             16 bytes     64 bytes    256 bytes   1024 bytes   8192 bytes  16384 bytes
AES-128-GCM     643091.50k  1275721.92k  3073367.81k  4366619.65k  4996866.05k  5057806.34k
get-info        179100.86k        0.00         0.00         0.00         0.00         0.00
```

Patched:
```
[mraz@leccinum ossl2]$ LD_LIBRARY_PATH=. apps/openssl speed -seconds 1 -evp aes-128-gcm
Doing AES-128-GCM for 1s on 16 size blocks: 40407762 AES-128-GCM's in 1.00s
Doing AES-128-GCM for 1s on 64 size blocks: 19804995 AES-128-GCM's in 1.00s
Doing AES-128-GCM for 1s on 256 size blocks: 11938205 AES-128-GCM's in 0.99s
Doing AES-128-GCM for 1s on 1024 size blocks: 4244186 AES-128-GCM's in 1.00s
Doing AES-128-GCM for 1s on 8192 size blocks: 604449 AES-128-GCM's in 1.00s
Doing AES-128-GCM for 1s on 16384 size blocks: 305132 AES-128-GCM's in 1.00s
Doing get-info for 1s on 16 size blocks: 26123854 get-info's in 1.00s
version: 3.1.0-dev
built on: Wed Dec  8 17:45:44 2021 UTC
options: bn(64,64)
compiler: gcc -fPIC -pthread -m64 -Wa,--noexecstack -Wall -O3 -DPEDANTIC -pedantic -Wno-long-long -DUNUSEDRESULT_DEBUG -Wall -Wmissing-declarations -Wextra
CPUINFO: OPENSSL_ia32cap=0x7ed8320b078bffff:0x400004219c91a9
The 'numbers' are in 1000s of bytes per second processed.
type             16 bytes     64 bytes    256 bytes   1024 bytes   8192 bytes  16384 bytes
AES-128-GCM     646524.19k  1267519.68k  3087050.99k  4346046.46k  4951646.21k  4999282.69k
get-info        417981.66k        0.00         0.00         0.00         0.00         0.00.
```

So the patched code is more than twice as fast.

**Review by paulidale — COMMENTED** [review]
I'm not sure I like the kludge which will be pervasive (there are a lot more places to instrument).

The case where `get_ctx_params()` calls `generic_get_ctx_params()` should also be covered, which requires passing the used count back.

There are also places where params come in groups (some of the TLS ones).  I don't think this complicates anything however.

**Comment by t8m** [comment]
I like the more generic ptrie approach more unfortunately when I cherry picked the openssl speed patch from this PR and applied it over your ptrie PR @paulidale, this was the result:
```
[mraz@leccinum ossl2]$ LD_LIBRARY_PATH=. apps/openssl speed -seconds 1 -evp aes-128-gcm
Doing AES-128-GCM for 1s on 16 size blocks: 39955098 AES-128-GCM's in 1.00s
Doing AES-128-GCM for 1s on 64 size blocks: 20472995 AES-128-GCM's in 1.00s
Doing AES-128-GCM for 1s on 256 size blocks: 11662145 AES-128-GCM's in 1.00s
Doing AES-128-GCM for 1s on 1024 size blocks: 4235636 AES-128-GCM's in 1.00s
Doing AES-128-GCM for 1s on 8192 size blocks: 607921 AES-128-GCM's in 1.00s
Doing AES-128-GCM for 1s on 16384 size blocks: 308065 AES-128-GCM's in 1.00s
Doing get-info for 1s on 16 size blocks: 9850801 get-info's in 1.00s
version: 3.1.0-dev
built on: Thu Dec 16 10:52:52 2021 UTC
options: bn(64,64)
compiler: gcc -fPIC -pthread -m64 -Wa,--noexecstack -Wall -O3 -DPEDANTIC -pedantic -Wno-long-long -DUNUSEDRESULT_DEBUG -Wall -Wmissing-declarations -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wswitch -Wsign-compare -Wshadow -Wformat -Wtype-limits -Wundef -Werror -Wmissing-prototypes -Wstrict-prototypes -g -O2 -Werror -DOPENSSL_USE_NODELETE -DL_ENDIAN -DOPENSSL_PIC -DOPENSSL_BUILDING_OPENSSL -DZLIB -DNDEBUG
CPUINFO: OPENSSL_ia32cap=0x7ed8320b078bffff:0x400004219c91a9
The 'numbers' are in 1000s of bytes per second processed.
type             16 bytes     64 bytes    256 bytes   1024 bytes   8192 bytes  16384 bytes
AES-128-GCM     639281.57k  1310271.68k  2985509.12k  4337291.26k  4980088.83k  5047336.96k
get-info        157612.82k        0.00         0.00         0.00         0.00         0.00 
```

**Comment by t8m** [comment]
So as you can see calling just EVP_CIPHER_CTX_get_key_length and EVP_CIPHER_CTX_get_iv_length is still slower than original unpatched code.

**Comment by paulidale** [comment]
We need to make `EVP_CIPHER_CTX_get_key_length` and `EVP_CIPHER_CTX_get_iv_length` calls caching.

**Comment by t8m** [comment]
> We need to make `EVP_CIPHER_CTX_get_key_length` and `EVP_CIPHER_CTX_get_iv_length` calls caching.

How do you do that? Also that helps only if ctx is reused. I am not sure if that was the case in the original issue that triggered this development.

**Comment by paulidale** [comment]
There was some reuse I believe.
Or repeated calls to these functions on the same context.

File: apps/speed.c [hunk 1]
```diff
@@ -750,6 +750,24 @@ static int EVP_Update_loop(void *args)
     return count;
 }
 
+static int EVP_Ciph_get_info_loop(void *args)
```
Comment (t-j-h) [inline apps/speed.c:753]: really don't like that function name ...

**Comment by mattcaswell** [comment]
OTC: Instead of this approach we should just cache the results of `EVP_CIPHER_CTX_get_key_length()` etc

**Pull request closed without merging**